File: backend/app/api/routes/shipments.py
"""运单接口：列表（分页+筛选）、详情（含运输段）、里程碑事件、轨迹查询。"""

from datetime import datetime

from fastapi import APIRouter, Depends, HTTPException, Query
from sqlalchemy import func, select
from sqlalchemy.orm import aliased

from app.api.deps import current_user
from app.api.schemas import (
    CarrierOption,
    LegOut,
    MilestoneEventOut,
    PagedShipments,
    PositionPointOut,
    ShipmentBrief,
    ShipmentDetail,
)
from app.db.models.reference import Carrier, Location
from app.db.models.shipment import Leg, Order, Shipment
from app.db.models.tracking import MilestoneEvent, PositionPoint
from app.db.session import SessionDep

router = APIRouter(prefix="/shipments", tags=["shipments"], dependencies=[Depends(current_user)])

# 起运地和目的地都指向 locations，用别名区分两次 join
Origin = aliased(Location, name="origin")
Dest = aliased(Location, name="dest")
# 运输段同样有起止地点，另建一对别名，避免与运单的那对冲突
LegOrigin = aliased(Location, name="leg_origin")
LegDest = aliased(Location, name="leg_dest")


@router.get("/carriers", response_model=list[CarrierOption])
async def list_carriers(session: SessionDep) -> list[CarrierOption]:
    """承运商下拉项：供分析页下钻筛选运单时渲染 Select。放在 /{shipment_id} 之前避免被路径参数吞掉。"""
    rows = (await session.execute(select(Carrier.id, Carrier.name, Carrier.mode))).all()
    return [CarrierOption(id=id, name=name, mode=mode) for id, name, mode in rows]


@router.get("", response_model=PagedShipments)
async def list_shipments(
    session: SessionDep,
    page: int = Query(1, ge=1),
    page_size: int = Query(20, ge=1, le=100),
    status: str | None = Query(None, description="planned / in_transit / delivered / delayed"),
    shipment_no: str | None = Query(None, description="按运单号模糊匹配（部分即可）"),
    carrier_id: int | None = Query(None, description="按承运商筛选，分析页点击承运商下钻时传入"),
) -> PagedShipments:
    """运单列表：分页 + 按状态/运单号/承运商筛选，带出港口 code 和承运商名。"""
    conditions = []
    if status:
        conditions.append(Shipment.status == status)
    if shipment_no:
        # 模糊匹配，用户输入部分运单号即可定位（运单号含固定前缀，前缀检索仍有意义）
        conditions.append(Shipment.shipment_no.ilike(f"%{shipment_no}%"))
    if carrier_id is not None:
        conditions.append(Shipment.carrier_id == carrier_id)

    total = await session.scalar(select(func.count()).select_from(Shipment).where(*conditions))

    stmt = (
        select(Shipment, Origin.code, Dest.code, Carrier.name)
        .outerjoin(Origin, Shipment.origin_id == Origin.id)
        .outerjoin(Dest, Shipment.dest_id == Dest.id)
        .outerjoin(Carrier, Shipment.carrier_id == Carrier.id)
        .where(*conditions)
        .order_by(Shipment.id.desc())
        .offset((page - 1) * page_size)
        .limit(page_size)
    )
    rows = (await session.execute(stmt)).all()

    items = [
        ShipmentBrief(
            id=s.id,
            shipment_no=s.shipment_no,
            status=s.status,
            origin_code=origin_code,
            dest_code=dest_code,
            carrier_name=carrier_name,
            planned_departure=s.planned_departure,
            planned_arrival=s.planned_arrival,
            actual_departure=s.actual_departure,
            actual_arrival=s.actual_arrival,
            latest_lat=s.latest_lat,
            latest_lng=s.latest_lng,
            latest_ts=s.latest_ts,
        )
        for s, origin_code, dest_code, carrier_name in rows
    ]

    return PagedShipments(total=total or 0, page=page, page_size=page_size, items=items)


@router.get("/{shipment_id}", response_model=ShipmentDetail)
async def get_shipment(shipment_id: int, session: SessionDep) -> ShipmentDetail:
    """运单详情：带出各运输段（legs）。"""
    stmt = (
        select(Shipment, Origin.code, Dest.code, Carrier.name, Order.order_no, Order.customer_name)
        .outerjoin(Origin, Shipment.origin_id == Origin.id)
        .outerjoin(Dest, Shipment.dest_id == Dest.id)
        .outerjoin(Carrier, Shipment.carrier_id == Carrier.id)
        .outerjoin(Order, Shipment.order_id == Order.id)
        .where(Shipment.id == shipment_id)
    )
    row = (await session.execute(stmt)).first()
    if row is None:
        raise HTTPException(status_code=404, detail="运单不存在")

    s, origin_code, dest_code, carrier_name, order_no, customer_name = row

    # 段也要 join locations 拿港口 code，否则前端链路只能显示地点 id
    leg_rows = (
        await session.execute(
            select(
                Leg,
                LegOrigin.code,
                LegDest.code,
                LegOrigin.lat,
                LegOrigin.lng,
                LegDest.lat,
                LegDest.lng,
            )
            .outerjoin(LegOrigin, Leg.origin_id == LegOrigin.id)
            .outerjoin(LegDest, Leg.dest_id == LegDest.id)
            .where(Leg.shipment_id == shipment_id)
            .order_by(Leg.seq)
        )
    ).all()

    # 总起 / 总止经纬度：第一段 origin 与最后一段 dest 的 lat/lng，
    # 用于前端地图画"完整规划路径"虚线（已走过的轨迹只到实时点为止，
    # 没有这些就看不到剩下还要走的路）。
    origin_lat = origin_lng = dest_lat = dest_lng = None
    if leg_rows:
        # leg_rows 已 join 出每段 origin/dest 的 lat/lng（参见上面的 select），
        # 直接取第一段 origin 与最后一段 dest，避免再单独查 Location。
        _, _, _, o_lat, o_lng, _, _ = leg_rows[0]
        _, _, _, _, _, d_lat, d_lng = leg_rows[-1]
        origin_lat, origin_lng = o_lat, o_lng
        dest_lat, dest_lng = d_lat, d_lng

    return ShipmentDetail(
        id=s.id,
        shipment_no=s.shipment_no,
        status=s.status,
        origin_code=origin_code,
        dest_code=dest_code,
        carrier_name=carrier_name,
        planned_departure=s.planned_departure,
        planned_arrival=s.planned_arrival,
        actual_departure=s.actual_departure,
        actual_arrival=s.actual_arrival,
        latest_lat=s.latest_lat,
        latest_lng=s.latest_lng,
        latest_ts=s.latest_ts,
        order_id=s.order_id,
        order_no=order_no,
        customer_name=customer_name,
        origin_lat=origin_lat,
        origin_lng=origin_lng,
        dest_lat=dest_lat,
        dest_lng=dest_lng,
        legs=[
            LegOut(
                id=leg.id,
                seq=leg.seq,
                mode=leg.mode,
                origin_id=leg.origin_id,
                dest_id=leg.dest_id,
                origin_code=leg_origin_code,
                dest_code=leg_dest_code,
                origin_lat=leg_origin_lat,
                origin_lng=leg_origin_lng,
                dest_lat=leg_dest_lat,
                dest_lng=leg_dest_lng,
                planned_start=leg.planned_start,
                planned_end=leg.planned_end,
                status=leg.status,
            )
            for leg, leg_origin_code, leg_dest_code, leg_origin_lat, leg_origin_lng, leg_dest_lat, leg_dest_lng in leg_rows
        ],
    )


@router.get("/{shipment_id}/events", response_model=list[MilestoneEventOut])
async def get_shipment_events(shipment_id: int, session: SessionDep) -> list[MilestoneEventOut]:
    """运单的里程碑事件时间轴。"""
    events = (
        await session.scalars(
            select(MilestoneEvent)
            .where(MilestoneEvent.shipment_id == shipment_id)
            .order_by(MilestoneEvent.occurred_at)
        )
    ).all()
    return [MilestoneEventOut.model_validate(e) for e in events]


@router.get("/{shipment_id}/positions", response_model=list[PositionPointOut])
async def get_shipment_positions(
    shipment_id: int,
    session: SessionDep,
    start: datetime | None = Query(None, description="起始时间（UTC）"),
    end: datetime | None = Query(None, description="结束时间（UTC）"),
    limit: int = Query(2000, ge=1, le=10000),
) -> list[PositionPointOut]:
    """运单的历史轨迹：先取该运单所有段，再按时间范围查轨迹点。"""
    leg_ids = (await session.scalars(select(Leg.id).where(Leg.shipment_id == shipment_id))).all()
    if not leg_ids:
        return []

    conditions = [PositionPoint.leg_id.in_(leg_ids)]
    if start:
        conditions.append(PositionPoint.recorded_at >= start)
    if end:
        conditions.append(PositionPoint.recorded_at <= end)

    stmt = select(PositionPoint).where(*conditions).order_by(PositionPoint.recorded_at).limit(limit)
    points = (await session.scalars(stmt)).all()
    return [PositionPointOut.model_validate(p) for p in points]

File: frontend/src/types/shipments.ts
// 运单领域 / 契约类型：对齐后端 app/api/routes/shipments.py 的真实返回结构。
// 作为无依赖的单一来源，api / constants / components 均从此处引用，
// 不反向依赖传输层（api），保持依赖方向单向、无环。

export type ShipmentStatus = 'planned' | 'in_transit' | 'delivered' | 'delayed'
export type TransportMode = 'sea' | 'air' | 'rail' | 'road'

/** 运单列表项：带起运/目的港 code 与承运商名，前端可直接展示。 */
export interface ShipmentBrief {
  id: number
  shipment_no: string
  status: string
  origin_code: string | null
  dest_code: string | null
  carrier_name: string | null
  planned_departure: string | null
  planned_arrival: string | null
  actual_departure: string | null
  actual_arrival: string | null
  latest_lat: number | null
  latest_lng: number | null
  latest_ts: string | null
}

/** 运输段：一票货的其中一段（港口用 id 关联，前端展示 mode/状态即可）。 */
export interface LegOut {
  id: number
  seq: number
  mode: string
  origin_id: number | null
  dest_id: number | null
  origin_code: string | null
  dest_code: string | null
  /** 起止港口经纬度：后端 join 带出，前端按段画规划路径基线。 */
  origin_lat: number | null
  origin_lng: number | null
  dest_lat: number | null
  dest_lng: number | null
  planned_start: string | null
  planned_end: string | null
  status: string
}

/** 运单详情：在列表项基础上带出各运输段。 */
export interface ShipmentDetail extends ShipmentBrief {
  order_id: number | null
  /** 订单号：后端 join orders 带出，页面展示这个而非裸 id */
  order_no: string | null
  /** 货主：同属订单信息，从 orders 带出 */
  customer_name: string | null
  legs: LegOut[]
  /** 总起经纬度（第一段 origin），用于画完整规划路径虚线。 */
  origin_lat: number | null
  origin_lng: number | null
  /** 总止经纬度（最后一段 dest），用于画完整规划路径虚线。 */
  dest_lat: number | null
  dest_lng: number | null
}

/** 里程碑事件（时间轴用）。 */
export interface MilestoneEventOut {
  id: number
  shipment_id: number
  leg_id: number | null
  event_type: string
  occurred_at: string
  payload_json: string | null
}

/** 轨迹点（地图用，带经纬度）。 */
export interface PositionPointOut {
  id: number
  leg_id: number
  lat: number
  lng: number
  speed: number | null
  heading: number | null
  recorded_at: string
}

/** 运单分页结果。 */
export interface PagedShipments {
  total: number
  page: number
  page_size: number
  items: ShipmentBrief[]
}

export interface ShipmentsQuery {
  page?: number
  page_size?: number
  status?: string | null
  /** 按运单号模糊匹配（前端列表搜索框）。 */
  shipment_no?: string | null
}

/** 全局事件筛选参数（通知中心）。 */
export interface EventsQuery {
  page?: number
  page_size?: number
  event_type?: string | null
  shipment_no?: string | null
  since?: string | null
}

/** 全局里程碑事件（通知中心列表项）。 */
export interface EventOut {
  id: number
  shipment_id: number
  leg_id: number | null
  event_type: string
  occurred_at: string
  payload_json: string | null
  shipment_no: string | null
  shipment_status: string | null
}

/** 事件分页结果。 */
export interface PagedEvents {
  total: number
  page: number
  page_size: number
  items: EventOut[]
}

/** 异常筛选参数（异常中心）。 */
export interface ExceptionsQuery {
  page?: number
  page_size?: number
  type?: string | null
  level?: string | null
  unresolved_only?: boolean
}

/** 异常记录（异常中心列表项）。 */
export interface ExceptionOut {
  id: number
  shipment_id: number
  type: string
  level: string
  detail: string | null
  detected_at: string
  resolved_at: string | null
  shipment_no: string | null
  shipment_status: string | null
}

/** 异常分页结果。 */
export interface PagedExceptions {
  total: number
  page: number
  page_size: number
  items: ExceptionOut[]
}

/** 大屏地图：地点标注（港口/仓库/城市统一）。 */
export interface MapPort {
  code: string
  name: string
  lat: number
  lng: number
}

/** 大屏地图：航线段，起止经纬度已由后端 join 带出。 */
export interface MapLeg {
  seq: number
  mode: string
  /** 实时位置流按 leg_id 上报，靠它把位置增量映射回运单与段 */
  leg_id: number
  origin_code: string | null
  dest_code: string | null
  origin_lat: number | null
  origin_lng: number | null
  dest_lat: number | null
  dest_lng: number | null
  status: string
  /** 活动段最近轨迹点（[lng, lat, ts_ms] 升序）：
   * 既是「实际轨迹」的历史垫底，也供时间轴回放按当前时刻切片；
   * ts_ms 为 UTC epoch 毫秒。 */
  track: [number, number, number][]
}

/** 大屏地图：运单航线，聚合各段供按运单着色与点击交互。 */
export interface MapRoute {
  shipment_id: number
  shipment_no: string
  status: string
  latest_lat: number | null
  latest_lng: number | null
  legs: MapLeg[]
}

/** 大屏地图聚合数据：地点 + 航线，一次请求拿全。 */
export interface MapOverview {
  ports: MapPort[]
  routes: MapRoute[]
}
